### DISCUSSION TITLE
What does the index of the embedding mean?

### DISCUSSION THREAD

**Discussion by vadhoob** [krishnanlab/PecanPy#138] [Q&A] [answered]
Hi, 

I am currently generating embedding using g.embed  method using the following code block: 

`g = node2vec.PreComp(p=in_parameter,
                                            q=out_parameter,
                                            workers=4,
                                            verbose=True)`
    `g.read_edg(edges_path,
                        delimiter=',',
                        weighted=False,
                       directed=False)`

   `g.preprocess_transition_probs()`

 ` emb = g.embed(dim=num_dim,
                               num_walks=number_of_walks,
                               walk_length=lenght_walk,
                               window_size=size_window,
                               epochs=num_epochs)`

I then use for example emb[0] to get the embedding associated with node 1. Is this the correct interpretation of what g.emd produces? I am using these embeddings for a downstream classification task where I join these embeddings with other features based on index key. 

Many thanks!

**Comment by RemyLau** [comment, marked as answer]
Hi @vadhoob, thanks for your interest in using PecanPy! In this case, the index of `emd` will be consistent with the graph `g`, i.e., the first row in the `emd` matrix correspond to the embedding vector of the first node in `g`.

The relevant code is as follows
https://github.com/krishnanlab/PecanPy/blob/a2b5f7df790e58c7cccdfcb19039640ee222b242/src/pecanpy/pecanpy.py#L285

where `w2v` is the trained [word2vec](https://radimrehurek.com/gensim/models/word2vec.html#module-gensim.models.word2vec) object, and `w2v.wv` is the corresponding [keyedvectors](https://radimrehurek.com/gensim/models/keyedvectors.html#gensim.models.keyedvectors.KeyedVectors). This keyedvectors can be indexed by the corresponding words (i.e., node IDs in the graph). Thus, using the node IDs `self.nodes` of the graph `g` via `w2v.wv[self.nodes]` produces an embedding matrix whose node order is consistent with the graph.

**Reply by vadhoob** [reply]
Thanks @RemyLau, based on your reply this is what I have done: 

`node_list = [int(x) for x in g.nodes]` 

and then:

`embedding_with_key = dict(zip(node_list, emb))`

This is now gives me a dictionary containing the node id and its corresponding embedding. Is this correct? I am trying to avoid doing any loop operation as my graph is quite big. 

Many thanks!

**Reply by RemyLau** [reply]
Yes, this looks correct to me. Maybe if you only need the `embedding_with_key` but not the node_list, you can simply do the following

```python
embedding_with_str_key = dict(zip(g.nodes, emb))

# Or map the nodes to int if that's what you wanted as well
embedding_with_int_key = dict(zip(map(int, g.nodes), emb))
```

**Reply by vadhoob** [reply]
okay thank you!
